In [1]:
import numpy as np

### Part 1. 1D Array indexing.

In [2]:
my_array = np.array([i for i in range(10)])

In [4]:
my_array

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [3]:
my_array[5]

5

In [5]:
my_array[-1]

9

In [6]:
my_array[3:6]

array([3, 4, 5])

In [9]:
my_array[1:8:3]

array([1, 4, 7])

### Part 2. ND Array indexing.

In [10]:
my_array = np.array([[1,2,3,4], [10,11,12,13], [45,46,47,48]])

In [11]:
my_array

array([[ 1,  2,  3,  4],
       [10, 11, 12, 13],
       [45, 46, 47, 48]])

In [12]:
my_array[1][2]

12

In [13]:
my_array[1, 2]

12

In [14]:
my_array[1]

array([10, 11, 12, 13])

In [15]:
my_array[:, 2]

array([ 3, 12, 47])

In [16]:
my_array[1:, 1:3]

array([[11, 12],
       [46, 47]])

### Part 3. Tasks.

In [17]:
first_line = [x*y for x in range(2, 100, 6) for y in range (7, 1, -2)]
second_line = [x ** 0.5 for x in range(1000, 1101, 2)]
third_line = [x**2 for x in range(51)]

big_secret = np.array([first_line, second_line, third_line, second_line, first_line])

In [18]:
big_secret

array([[1.40000000e+01, 1.00000000e+01, 6.00000000e+00, 5.60000000e+01,
        4.00000000e+01, 2.40000000e+01, 9.80000000e+01, 7.00000000e+01,
        4.20000000e+01, 1.40000000e+02, 1.00000000e+02, 6.00000000e+01,
        1.82000000e+02, 1.30000000e+02, 7.80000000e+01, 2.24000000e+02,
        1.60000000e+02, 9.60000000e+01, 2.66000000e+02, 1.90000000e+02,
        1.14000000e+02, 3.08000000e+02, 2.20000000e+02, 1.32000000e+02,
        3.50000000e+02, 2.50000000e+02, 1.50000000e+02, 3.92000000e+02,
        2.80000000e+02, 1.68000000e+02, 4.34000000e+02, 3.10000000e+02,
        1.86000000e+02, 4.76000000e+02, 3.40000000e+02, 2.04000000e+02,
        5.18000000e+02, 3.70000000e+02, 2.22000000e+02, 5.60000000e+02,
        4.00000000e+02, 2.40000000e+02, 6.02000000e+02, 4.30000000e+02,
        2.58000000e+02, 6.44000000e+02, 4.60000000e+02, 2.76000000e+02,
        6.86000000e+02, 4.90000000e+02, 2.94000000e+02],
       [3.16227766e+01, 3.16543836e+01, 3.16859590e+01, 3.17175031e+01,
       

In [22]:
big_secret.shape

(5, 51)

#### Task 1.
Чему равна сумма элементов последнего столбца массива? Ответ округлите до двух цифр после запятой.

#### Solution:

In [27]:
col_sum = 0
for item in big_secret[:, -1]:
    col_sum += item

print(f'Last column elements sum: {col_sum:.2f}')

Last column elements sum: 3154.33


#### Conclusion: 3154.33 is the right answer!!!

#### Task 2.
Выделите из каждой строки массива big_secret первые 5 элементов. Чему равна сумма элементов главной диагонали получившейся матрицы? Округлите ответ до двух цифр после запятой.

#### Solution:

In [29]:
temp_array = big_secret[:, :5]

diag_sum = 0
for i in range(5):
    diag_sum += temp_array[i, i]

print(f'Main diagonal elements sum: {diag_sum:.2f}')

Main diagonal elements sum: 121.37


#### Conclusion: 121.37 is the right answer!!!

#### Task 3.
Выделите из каждой строки массива big_secret последние 5 элементов. Чему равно произведение элементов главной диагонали получившейся матрицы? Введите полученный результат без изменений и округлений.

#### Solution:

In [33]:
temp_array = big_secret[:, -5:]
display(temp_array)

diag_mul = 1
for i in range(5):
    diag_mul *= temp_array[i, i]

print(f'Main diagonal elements mul: {diag_mul}')

array([[ 460.        ,  276.        ,  686.        ,  490.        ,
         294.        ],
       [  33.04542328,   33.07567082,   33.10589071,   33.13608305,
          33.1662479 ],
       [2116.        , 2209.        , 2304.        , 2401.        ,
        2500.        ],
       [  33.04542328,   33.07567082,   33.10589071,   33.13608305,
          33.1662479 ],
       [ 460.        ,  276.        ,  686.        ,  490.        ,
         294.        ]])

Main diagonal elements mul: 341505315559.2347


#### Conclusion: 341505315559.2347 is the right answer!!!

#### Common operations:
Проверим знания! Продолжим работать с массивом big_secret. Замените на 1 все элементы, у которых оба индекса нечётные, и на -1 все элементы, у которых оба индекса чётные.

In [40]:
(rows, cols) = big_secret.shape

for row in range(rows):
    for col in range(cols):
        if row%2 and col%2:
            big_secret[row, col] = 1
        elif (not row%2) and (not col%2):
            big_secret[row, col] = -1

display(big_secret)

array([[-1.00000000e+00,  1.00000000e+01, -1.00000000e+00,
         5.60000000e+01, -1.00000000e+00,  2.40000000e+01,
        -1.00000000e+00,  7.00000000e+01, -1.00000000e+00,
         1.40000000e+02, -1.00000000e+00,  6.00000000e+01,
        -1.00000000e+00,  1.30000000e+02, -1.00000000e+00,
         2.24000000e+02, -1.00000000e+00,  9.60000000e+01,
        -1.00000000e+00,  1.90000000e+02, -1.00000000e+00,
         3.08000000e+02, -1.00000000e+00,  1.32000000e+02,
        -1.00000000e+00,  2.50000000e+02, -1.00000000e+00,
         3.92000000e+02, -1.00000000e+00,  1.68000000e+02,
        -1.00000000e+00,  3.10000000e+02, -1.00000000e+00,
         4.76000000e+02, -1.00000000e+00,  2.04000000e+02,
        -1.00000000e+00,  3.70000000e+02, -1.00000000e+00,
         5.60000000e+02, -1.00000000e+00,  2.40000000e+02,
        -1.00000000e+00,  4.30000000e+02, -1.00000000e+00,
         6.44000000e+02, -1.00000000e+00,  2.76000000e+02,
        -1.00000000e+00,  4.90000000e+02, -1.00000000e+0

#### Task 4.
Выделите из каждой строки обновлённого массива big_secret первые 5 элементов. Чему равна сумма элементов главной диагонали получившейся матрицы? Введите полученный ответ без изменений и округлений.

#### Solution:

In [41]:
temp_array = big_secret[:, :5]
display(temp_array)

diag_sum = 0
for i in range(5):
    diag_sum += temp_array[i, i]

print(f'Main diagonal elements sum: {diag_sum}')

array([[-1.        , 10.        , -1.        , 56.        , -1.        ],
       [31.6227766 ,  1.        , 31.68595904,  1.        , 31.74901573],
       [-1.        ,  1.        , -1.        ,  9.        , -1.        ],
       [31.6227766 ,  1.        , 31.68595904,  1.        , 31.74901573],
       [-1.        , 10.        , -1.        , 56.        , -1.        ]])

Main diagonal elements sum: -1.0


#### Conclusion: -1 is the right answer!!!

#### Task 5.
Выделите из каждой строки обновлённого массива big_secret последние 5 элементов. Чему равно произведение элементов главной диагонали получившейся матрицы? Введите полученный результат без изменений и округлений.

#### Solution:

In [42]:
temp_array = big_secret[:, :5]
display(temp_array)

diag_mul = 1
for i in range(5):
    diag_mul *= temp_array[i, i]

print(f'Main diagonal elements mul: {diag_mul}')

array([[-1.        , 10.        , -1.        , 56.        , -1.        ],
       [31.6227766 ,  1.        , 31.68595904,  1.        , 31.74901573],
       [-1.        ,  1.        , -1.        ,  9.        , -1.        ],
       [31.6227766 ,  1.        , 31.68595904,  1.        , 31.74901573],
       [-1.        , 10.        , -1.        , 56.        , -1.        ]])

Main diagonal elements mul: -1.0


#### Conclusion: -1 is the right answer!!!